tests/test_algorithm_learning.py

```diff
@@ -0,0 +1,355 @@
+        self._episode += 1
+        if self._is_valid(action):
+            self._flip_state()
+            return self._current_state, self._reward.sample(), self._is_done(), None
+        else:
+            return self._current_state, self._penalty.sample(), self._is_done(), None
+
+    def _is_valid(self, action):
+        return action is not None and self._current_state[0] != action
+
+    def _flip_state(self):
+        self._current_state = np.array([self._current_state[1], self._current_state[0]])
+
+    def _is_done(self):
+        return self._episode == self.len_episode
+
+    def __repr__(self):
+        return f"SimpleDeterministic(action_space={repr(self.action_space)}, " \
+               f"observation_space={repr(self.observation_space)}, len_episode={self.len_episode}," \
+               f"avg_total_return={repr(self.avg_max_reward)})"
+
+
+class RandomAgent(AgentInterface):
+    def __init__(self, action_space):
+        self._action_space = action_space
+
+    def next_action(self, _):
+        return self._action_space.sample()
+
+    def signal(self, _):
+        pass
+
+    def train(self):
+        pass
+
+
+class OptimalAgent(AgentInterface):
+    def next_action(self, observation):
+        return 1 - observation[0]
+
+    def signal(self, reward):
+        pass
+
+    def train(self):
+        pass
+
+
+class LinearPolicy:
+    def __init__(self, plotter, lr=1):
+        self._lr = lr
+        self._m = np.random.normal(1, 0.001)  # high m avoids local maxima - use negative m for demonstration
+        self._probs = []
+        self._get_signal = None
+        self._plotter = plotter(self._m)
+
+    def set_signal_calc(self, signal_calc):
+        def np_signal(acts, ps, rs):
+            return signal_calc(np_ops, acts, ps, rs)
+
+        self._get_signal = np_signal
+
+    def estimate(self, observation):
+        y = (np.tanh(self._m * ((observation[0] * 2) - 1)) + 1) / 2
+        p = np.array([y, 1 - y], dtype=np.float)
+        self._probs.append(p)
+        return p
+
+    def fit(self, trajectory):
+        signal = self._get_signal(trajectory.actions, np.array(self._probs, dtype=np.float), trajectory.returns)
+        self._m += signal * self._lr
+        self._probs.clear()
+        self._plotter.record(self._m, trajectory.returns, signal)
+
+    def plot(self):
+        self._plotter.show()
+
+    class MatPlotter:
+        def __init__(self, m):
+            self._m_history = [m]
+            self._ret_med_history = [0.0]
+            self._signal_history = [0]
+
+        def record(self, m, returns, signal):
+            self._m_history.append(m)
+            self._signal_history.append(signal)
+            self._ret_med_history.append(float(np.mean(returns)))
+
+        def show(self):
+            import matplotlib.pyplot as plt
+            fig, (mp, lp, rp) = plt.subplots(3, 1)
+            fig.suptitle("Numpy Agent")
+            mp.plot(self._m_history)
+            mp.set_title("m")
```

Comment:
```suggestion
            mp.set_title("slope")
```
same here for the string